File: packages/agent-mesh/sdks/go/client_test.go
package agentmesh

import (
	"testing"
)

func TestNewClient(t *testing.T) {
	client, err := NewClient("test-agent",
		WithCapabilities([]string{"read", "write"}),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	if client.Identity == nil {
		t.Fatal("Identity is nil")
	}
	if client.Trust == nil {
		t.Fatal("Trust is nil")
	}
	if client.Policy == nil {
		t.Fatal("Policy is nil")
	}
	if client.Audit == nil {
		t.Fatal("Audit is nil")
	}
}

func TestExecuteWithGovernanceAllow(t *testing.T) {
	client, _ := NewClient("gov-agent",
		WithPolicyRules([]PolicyRule{
			{Action: "data.read", Effect: Allow},
		}),
	)

	result, err := client.ExecuteWithGovernance("data.read", nil)
	if err != nil {
		t.Fatalf("ExecuteWithGovernance: %v", err)
	}
	if !result.Allowed {
		t.Error("expected Allowed = true")
	}
	if result.Decision != Allow {
		t.Errorf("decision = %q, want allow", result.Decision)
	}
	if result.AuditEntry == nil {
		t.Error("expected AuditEntry to be non-nil")
	}
}

func TestExecuteWithGovernanceDeny(t *testing.T) {
	client, _ := NewClient("gov-agent",
		WithPolicyRules([]PolicyRule{
			{Action: "data.delete", Effect: Deny},
		}),
	)

	result, err := client.ExecuteWithGovernance("data.delete", nil)
	if err != nil {
		t.Fatal(err)
	}
	if result.Allowed {
		t.Error("expected Allowed = false for denied action")
	}
}

func TestEndToEndGovernance(t *testing.T) {
	client, _ := NewClient("e2e-agent",
		WithCapabilities([]string{"data.read"}),
		WithPolicyRules([]PolicyRule{
			{Action: "data.read", Effect: Allow},
			{Action: "*", Effect: Deny},
		}),
	)

	// Allowed action
	r1, _ := client.ExecuteWithGovernance("data.read", nil)
	if !r1.Allowed {
		t.Error("data.read should be allowed")
	}

	// Denied action
	r2, _ := client.ExecuteWithGovernance("system.shutdown", nil)
	if r2.Allowed {
		t.Error("system.shutdown should be denied")
	}

	// Audit chain intact
	if !client.Audit.Verify() {
		t.Error("audit chain should verify")
	}

	// Trust score updated
	score := client.Trust.GetTrustScore(client.Identity.DID)
	if score.Overall == 0 {
		t.Error("trust score should be non-zero after interactions")
	}
}

File: packages/agent-mesh/sdks/typescript/tests/client.test.ts
import { AgentMeshClient } from '../src/client';

describe('AgentMeshClient', () => {
  let client: AgentMeshClient;

  beforeEach(() => {
    client = AgentMeshClient.create('test-agent', {
      capabilities: ['read', 'write'],
      policyRules: [
        { action: 'data.read', effect: 'allow' },
        { action: 'data.write', effect: 'allow', conditions: { role: 'admin' } },
        { action: 'data.delete', effect: 'deny' },
      ],
    });
  });

  describe('create()', () => {
    it('creates a client with a valid identity', () => {
      expect(client.identity.did).toMatch(/^did:agentmesh:test-agent:/);
    });

    it('exposes trust, policy, and audit subsystems', () => {
      expect(client.trust).toBeDefined();
      expect(client.policy).toBeDefined();
      expect(client.audit).toBeDefined();
    });
  });

  describe('executeWithGovernance()', () => {
    it('allows permitted actions', async () => {
      const result = await client.executeWithGovernance('data.read');
      expect(result.decision).toBe('allow');
      expect(result.trustScore).toBeDefined();
      expect(result.auditEntry).toBeDefined();
      expect(result.executionTime).toBeGreaterThanOrEqual(0);
    });

    it('denies forbidden actions', async () => {
      const result = await client.executeWithGovernance('data.delete');
      expect(result.decision).toBe('deny');
    });

    it('evaluates conditions', async () => {
      const allowed = await client.executeWithGovernance('data.write', { role: 'admin' });
      expect(allowed.decision).toBe('allow');

      const denied = await client.executeWithGovernance('data.write', { role: 'user' });
      expect(denied.decision).toBe('deny');
    });

    it('creates an audit entry for each action', async () => {
      await client.executeWithGovernance('data.read');
      await client.executeWithGovernance('data.delete');

      const entries = client.audit.getEntries();
      expect(entries).toHaveLength(2);
      expect(client.audit.verify()).toBe(true);
    });

    it('updates trust score after actions', async () => {
      const before = client.trust.getTrustScore(client.identity.did);

      await client.executeWithGovernance('data.read'); // allow → success
      const afterSuccess = client.trust.getTrustScore(client.identity.did);
      expect(afterSuccess.overall).toBeGreaterThanOrEqual(before.overall);

      await client.executeWithGovernance('data.delete'); // deny → failure
      const afterFailure = client.trust.getTrustScore(client.identity.did);
      expect(afterFailure.overall).toBeLessThan(afterSuccess.overall);
    });

    it('returns execution time', async () => {
      const result = await client.executeWithGovernance('data.read');
      expect(typeof result.executionTime).toBe('number');
    });
  });

  describe('end-to-end governance flow', () => {
    it('maintains audit chain integrity across multiple operations', async () => {
      const actions = ['data.read', 'data.delete', 'data.read', 'unknown.action'];

      for (const action of actions) {
        await client.executeWithGovernance(action);
      }

      expect(client.audit.verify()).toBe(true);
      expect(client.audit.getEntries()).toHaveLength(4);
    });

    it('applies default-deny for unknown actions', async () => {
      const result = await client.executeWithGovernance('unregistered.action');
      expect(result.decision).toBe('deny');
    });
  });
});
